### PR TITLE
Add slim connector description

### PR DESCRIPTION
**Pull request by hagen-danswer**
## Description
This should be used as a reference when adding slim connectors to existing connectors! 

Included are:
- example slim connector implementation
- example test format
- some comments on the pr itself
- updated documentation to reference this

### CODE REVIEW

File: .github/workflows/pr-python-connector-tests.yml [hunk 1]
```diff
@@ -24,6 +24,8 @@ env:
   GOOGLE_DRIVE_OAUTH_CREDENTIALS_JSON_STR: ${{ secrets.GOOGLE_DRIVE_OAUTH_CREDENTIALS_JSON_STR }}
   GOOGLE_GMAIL_SERVICE_ACCOUNT_JSON_STR: ${{ secrets.GOOGLE_GMAIL_SERVICE_ACCOUNT_JSON_STR }}
   GOOGLE_GMAIL_OAUTH_CREDENTIALS_JSON_STR: ${{ secrets.GOOGLE_GMAIL_OAUTH_CREDENTIALS_JSON_STR }}
+  # Slab
+  SLAB_BOT_TOKEN: ${{ secrets.SLAB_BOT_TOKEN }}
```
Comment (hagen-danswer) [inline .github/workflows/pr-python-connector-tests.yml:28]: ^ This is required to make the tests run in github.

When submitting the PR for merging, make sure to share this with one of the developers @ danswer

File: backend/danswer/connectors/README.md
Comment (hagen-danswer) [inline backend/danswer/connectors/README.md:1]: All changes to this file are just for contributors and do not need to be replicated for other slim connectors

File: backend/danswer/connectors/README.md [hunk 2]
```diff
@@ -11,11 +11,16 @@ Connectors come in 3 different flows:
...
   - Bulk indexes documents to reflect a point in time. This type of connector generally works by either pulling all
   documents via a connector's API or loads the documents from some sort of a dump file.
-- Poll connector:
+- Poll Connector:
   - Incrementally updates documents based on a provided time range. It is used by the background job to pull the latest
   changes and additions since the last round of polling. This connector helps keep the document index up to date
   without needing to fetch/embed/index every document which would be too slow to do frequently on large sets of
   documents.
+- Slim Connector:
+  - The purpose of this connector is to be a lighter weight method of checking all documents in the source to see if they still exist.
```
Comment (pablodanswer) [inline backend/danswer/connectors/README.md:20]: Nit: would reword slightly "The purpose of this connector is to be a"

File: backend/tests/daily/connectors/slab/test_slab_data.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (hagen-danswer) [inline backend/tests/daily/connectors/slab/test_slab_data.json:1]: We keep long texts to check against in another file as a json for code cleanliness

File: backend/danswer/connectors/README.md [hunk 4]
```diff
@@ -11,11 +11,16 @@ Connectors come in 3 different flows:
...
+  - This connector should be identical to the Poll or Load Connector except that it only fetches the IDs of the documents, not the documents themselves.
+  - This is used for our pruning job that removes old documents from the index.
```
Comment (pablodanswer) [inline backend/danswer/connectors/README.md:22]: "in/by our pruning job which"

File: backend/tests/daily/connectors/slab/test_slab_connector.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    )
+    return connector
+
+
+@pytest.mark.xfail(
+    reason=(
+        "Need a test account with a slab subscription to run this test."
+        "Trial only lasts 14 days."
+    )
+)
```
Comment (hagen-danswer) [inline backend/tests/daily/connectors/slab/test_slab_connector.py:37]: If there is not already a test verifying the content of at least one of the retrieved documents, please add one!

File: backend/tests/daily/connectors/slab/test_slab_connector.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    assert section.text.replace("\u2019", "'") == desired_test_data["section_text"]
+    assert section.link == desired_test_data["link"]
+
+
+@pytest.mark.xfail(
+    reason=(
+        "Need a test account with a slab subscription to run this test."
+        "Trial only lasts 14 days."
+    )
+)
```
Comment (hagen-danswer) [inline backend/tests/daily/connectors/slab/test_slab_connector.py:75]: This test just checks that the ids retrieved through the load_from_state() is a subset of the ids retrieved from retrieve_all_slim_documents().

We dont check direct equality between the 2 sets because there are some circumstances where the slim connector might find an id that is filtered out by load/poll connector. 

This is a case that is handled downstream and is therefor okay

File: backend/tests/daily/connectors/slab/test_slab_connector.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import time
+from pathlib import Path
+
+import pytest
+
+from danswer.configs.constants import DocumentSource
+from danswer.connectors.models import Document
+from danswer.connectors.slab.connector import SlabConnector
+
+
```
Comment (hagen-danswer) [inline backend/tests/daily/connectors/slab/test_slab_connector.py:12]: This is only necessary if loading data from an outside folder

File: backend/danswer/connectors/slab/connector.py [hunk 8]
```diff
@@ -28,6 +31,8 @@
 SLAB_GRAPHQL_MAX_TRIES = 10
 SLAB_API_URL = "https://api.slab.com/v1/graphql"
 
+_SLIM_BATCH_SIZE = 1000
+
```
Comment (hagen-danswer) [inline backend/danswer/connectors/slab/connector.py:35]: ^1000 is typical for this

File: backend/danswer/connectors/slab/connector.py [hunk 9]
```diff
@@ -158,21 +163,26 @@ def get_slab_url_from_title_id(base_url: str, title: str, page_id: str) -> str:
...
         self.batch_size = batch_size
-        self.slab_bot_token = slab_bot_token
+        self._slab_bot_token: str | None = None
 
     def load_credentials(self, credentials: dict[str, Any]) -> dict[str, Any] | None:
-        self.slab_bot_token = credentials["slab_bot_token"]
+        self._slab_bot_token = credentials["slab_bot_token"]
         return None
 
+    @property
```
Comment (hagen-danswer) [inline backend/danswer/connectors/slab/connector.py:180]: this was added for cleanliness. However, its specific to this connector so can be ignored

**Review by pablodanswer — APPROVED** [review]

File: backend/danswer/connectors/slab/connector.py [hunk 10]
```diff
@@ -227,3 +237,21 @@ def poll_source(
...
             time_filter=lambda t: start_time <= t <= end_time
         )
+
+    def retrieve_all_slim_documents(
+        self,
+        start: SecondsSinceUnixEpoch | None = None,
+        end: SecondsSinceUnixEpoch | None = None,
+    ) -> GenerateSlimDocumentOutput:
+        slim_doc_batch: list[SlimDocument] = []
+        for post_id in get_all_post_ids(self.slab_bot_token):
```
Comment (hagen-danswer) [inline backend/danswer/connectors/slab/connector.py:247]: Notice here we call only get_all_post_ids(self.slab_bot_token). 

This means when running this connector, we never retrieve all the additional information that load and poll connectors do retrieve, meaning this is a much lighter/faster process.

make sure: 
- the id creation process is identical 
- that you are NOT retrieving the rest of the data that is used to make the document
